h2/connection.py

```diff
@@ -1646,7 +1646,19 @@ def _receive_data_frame(self, frame):
         self._inbound_flow_control_window_manager.window_consumed(
             flow_controlled_length
         )
-        stream = self._get_stream_by_id(frame.stream_id)
+
+        try:
+            stream = self._get_stream_by_id(frame.stream_id)
+        except StreamClosedError:
+            # This stream is already closed - and yet we received a DATA frame.
+            # The received DATA frame counts towards the connection flow window.
+            # We need to manually to acknowledge the DATA frame to update the flow window of the connection.
+            # Otherwise the whole connection stalls due to the inbound flow window being 0.
+            if flow_controlled_length > 0:
+                self.increment_flow_control_window(flow_controlled_length)
```

Comment:
`increment_flow_control_window()` method has a side-effect within and I don't think that it is appropriate to use it here. `_receive_data_frame` method is intended to **return** frames and events, but now it can also send frames on it's own.

Reply:
We are already using that - in the existing flow when a DATA frame on a closed stream is received, it will emit a RST_FRAME to inform the other side that it should no longer send frames for this stream: https://github.com/python-hyper/hyper-h2/blob/306609aa72c5ad4860a72fc4caa5db9ef5ab5b26/h2/connection.py#L1492 and notice the `_prepare_for_sending` call in both cases.

Reply:
All `_receive_*_frame()` methods don't call `_prepare_for_sending()` method to send frames (I hope 😅), they return frames as a list instead. So only `_receive_frame()` method calls `_prepare_for_sending()` method. I think it would be nice to keep this simplicity, it will help to reason about this code's behaviour later.

Reply:
Yes - this makes totally sense - let me see how I can change my PR.